src/bin/pgcopydb/ld_transform.c

```diff
@@ -2609,3 +2669,297 @@ LogicalMessageValueEq(LogicalMessageValue *a, LogicalMessageValue *b)
 	/* makes compiler happy */
 	return false;
 }
+
+
+/*
+ * Identifiers such as schema, table, column comes from various
+ * sources(e.g. wal2json, test_decoding and source catalog) and some of them
+ * already escapes identifiers and few don't.
+ * We need to check if the identifier is already quoted or not before
+ * escaping it.
+ * Whatever we are here is not a fool proof escaping mechanism, but a best
+ * effort to make sure that the identifiers are normalized by quoting them
+ * if it is not already quoted.
+ *
+ * Here is an example:
+ * foo -> "foo"
+ * "foo" -> "foo"
+ * foo"bar -> "foo"bar"
+ * "foo -> ""foo"
+ *
+ * The goal of this normalization is to make sure that the identifiers are
+ * comparable in the context of Hash Table.
+ */
+#define NORMALIZED_PG_NAMEDATA_COPY(dst, src) \
+	{ \
+		int len = strlen(src); \
+		if (src[0] == '"' && src[len - 1] == '"') \
+		{ \
+			strlcpy(dst, src, PG_NAMEDATALEN); \
+		} \
+		else \
+		{ \
+			sformat(dst, PG_NAMEDATALEN, "\"%s\"", src); \
+		} \
+	}
+
+/*
+ * fetchGeneratedColumnsForTable fetches the generated columns set for the given
+ * table "nspname.relname".
+ *
+ * Returns true if the table has generated columns and "columns" is set to the
+ * generated columns set, false otherwise.
+ *
+ * NOTE: There is no error condition, if the cache is NULL, it means that we
+ * don't have any generated columns in the catalog.
+ */
+static bool
+fetchGeneratedColumnsForTable(GeneratedColumnsCache *cache,
+							  const char *nspname,
+							  const char *relname,
+							  GeneratedColumnSet **columns)
+{
+	/*
+	 * NULL cache means that we don't have any generated columns in the
+	 * catalog.
+	 */
+	if (cache == NULL)
+	{
+		return false;
+	}
+
+	GeneratedColumnsCache *item = NULL;
+
+	GeneratedColumnsCache_Lookup key = { 0 };
+
+	NORMALIZED_PG_NAMEDATA_COPY(key.nspname, nspname);
+	NORMALIZED_PG_NAMEDATA_COPY(key.relname, relname);
+
+	HASH_FIND(hh, cache, &key, sizeof(GeneratedColumnsCache_Lookup), item);
+
+	if (item == NULL)
+	{
+		return false;
+	}
+
+	if (item->columns == NULL)
+	{
+		log_error("BUG: Table \"%s.%s\" is in the cache but columns are NULL",
+				  nspname, relname);
+		return false;
+	}
+
+	*columns = item->columns;
+
+	log_trace("Table \"%s.%s\" has generated columns", nspname, relname);
+
+	return true;
+}
+
+
+/*
+ * isGeneratedColumn checks whether the given "attname" is a generated column.
+ *
+ * Returns true if the column is generated, false otherwise.
+ *
+ * NOTE: There is no error condition, if the cache is NULL, it means that we
+ * don't have any generated columns in the catalog.
```

Comment:
Do we still need to do a hash lookup if the cache (`*columns`) is null here? is it safe to use the API that way?

Reply:
Fixed it.